refactor: migrate expandable card to chakra

jb-0 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Migrates the expandable card to Chakra. Not 100% sure on the desired animation approach here, but I've gone with Chakra Fade - happy to revisit at a later date.

<!--- Describe your changes in detail -->

## Related Issue
#6374 
<!--- This project accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Screenshots
![image](https://user-images.githubusercontent.com/39060614/201292571-9d303621-934c-4106-a3d8-a5ceb3b62e71.png)
![image](https://user-images.githubusercontent.com/39060614/201292657-2a04d2d9-6720-42e7-a05a-4b8a04d33312.png)
![image](https://user-images.githubusercontent.com/39060614/201292728-82af0ebd-a0f8-46c6-a96b-60f0cdd3ab8d.png)

TylerAPfledderer reviewed (changes requested):
Hey @jb-0!

Indeed, I agree the `Accordion` multi-component is best approach here. However, it was not intended to have a component like the `Fade` component wrap `AccordionPanel`. Based on the original components motion elements, it should actually be the other way around because `AccordionPanel` uses Chakra's `Collapse` component.

At the moment, I think it would get a little messy to litter the diff with comments versus stating all of my thoughts in this one comment, but this component can take advantage of `Accordion`'s unique feature to greatly reduce the code.

I would remove the need of `isVisible` custom state in favor of a render prop to obtain [AccordionItem internal state](https://chakra-ui.com/docs/components/accordion/usage#accessing-the-internal-state).

The `allowToggle` boolean prop applied to `Accordion` will also keep the toggling internal as there are no external triggers to make it a controlled component. In addition, the `AccordionButton` [per Chakra suggestion](https://chakra-ui.com/docs/components/accordion/usage#import) should be wrapped in an element with the `heading` role (if not a heading element) to support [Accessibility Patterns](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/).

There are also a couple of elements that were used in the original component that are no longer needed in `Accordion`, such as wrapping a "button-like" around `less / more` vs just using `Box` because `AccordionButton` is one whole selectable area. You can run the `trackCustomEvent` function through the render prop using the `isExpandable` internal state. 

My only other comment is if this component should have it's own `Accordion` and not make the `AccordionItem` component the parent, and use the `Accordion` component at the parent level? It should have no change in behavior aside from also using the `allowMultiple` prop. But this is a matter of preference and the component should work either way. 

I hope this helps in improving this migration!😄

jb-0 commented:
Hi Tyler, thanks for the comment & Chakra insights. I've kept my replies inline to keep this somewhat trackable, committed some changes based on the comment.

> Hey @jb-0!
> 
> Indeed, I agree the `Accordion` multi-component is best approach here. However, it was not intended to have a component like the `Fade` component wrap `AccordionPanel`. Based on the original components motion elements, it should actually be the other way around because `AccordionPanel` uses Chakra's `Collapse` component.
>
The Chakra Accordion transition is a bit abrupt compared to the original component. So attempting to keep this as close to the original component as possible, which displayed the text via a fade. I've moved the fade to within the panel, but visually moving it down a level doesn't make a difference as far as I can tell?

> At the moment, I think it would get a little messy to litter the diff with comments versus stating all of my thoughts in this one comment, but this component can take advantage of `Accordion`'s unique feature to greatly reduce the code.
> 
> I would remove the need of `isVisible` custom state in favor of a render prop to obtain [AccordionItem internal state](https://chakra-ui.com/docs/components/accordion/usage#accessing-the-internal-state).
> 
> The `allowToggle` boolean prop applied to `Accordion` will also keep the toggling internal as there are no external triggers to make it a controlled component.

The render prop is interesting! I always find it wierd seeing function syntax within jsx like that, but it works for the most part. Unfortunately it means that the content can no longer be clicked to collapse. As we are doing a like for like component refactor I think this is quite important.

So instead I've opted to make it a controlled component through the use of the [index prop](https://github.com/chakra-ui/chakra-ui/discussions/2619). This removes the need for the `isVisible` conditional around the panel.

> In addition, the `AccordionButton` [per Chakra suggestion](https://chakra-ui.com/docs/components/accordion/usage#import) should be wrapped in an element with the `heading` role (if not a heading element) to support [Accessibility Patterns](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/).
>
Done


> There are also a couple of elements that were used in the original component that are no longer needed in `Accordion`, such as wrapping a "button-like" around `less / more` vs just using `Box` because `AccordionButton` is one whole selectable area. You can run the `trackCustomEvent` function through the render prop using the `isExpandable` internal state, triggering it if this state is false as it shown in the click handler for the parent element.
I've switched the more / less button (that was just a div) to Chakra Text now. As for the rest of the comment, I believe it's addressed by making the component controlled.

> My only other comment is if this component should have it's own `Accordion` and not make the `AccordionItem` component the parent, and use the `Accordion` component at the parent level? It should have no change in behavior aside from also using the `allowMultiple` prop. But this is a matter of preference and the component should work either way.
> 
That approach isn't permitted in Chakra, because of the way Chakra passes state between components internally, so an `AccordionItem` is required - see screenshot ![image](https://user-images.githubusercontent.com/39060614/201468210-26a4b2e9-4ee1-4f42-9651-2273c9227e88.png)


> I hope this helps in improving this migration!smile

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -1,93 +1,22 @@
+  AccordionItem,
+  AccordionPanel,
+  Box,
+  Button,
```
Cleanup unsed deps

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+    <Accordion
+      border="1px solid"
+      borderColor="border"
+      borderRadius="2px"
```
```suggestion
      borderRadius="sm"
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+      borderRadius="2px"
+      display="flex"
+      flex-direction="column"
+      marginBottom="1rem"
```
```suggestion
      marginBottom="4"
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
-            animate={isVisible ? "expanded" : "collapsed"}
-            initial={false}
+      <AccordionItem borderTopWidth="0" flex="1">
+        <h3 style={{ margin: 0 }}>
```
```suggestion
        <Heading as="h3" m={0}>
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+        <h3 style={{ margin: 0 }}>
+          <AccordionButton
+            width="100%"
+            padding={`1.5rem 1.5rem ${isVisible ? "0" : "1.5rem"}`}
```
Lets keep this padding for the entire box and remove the extra bottom margin from the title section (see the following comment)
```suggestion
            p={6}
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+                    fontSize="xl"
+                    fontWeight="semibold"
+                    flex="1"
+                    marginTop="0rem"
+                    marginBottom="0.5rem"
```
```suggestion
                    m="0"
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+          paddingTop="0"
+          onClick={onClick}
+        >
+          <Fade in={isVisible}>
```
Appreciate the effort to keep the same type of animation we had but I think that in this transition we can start using the default animation from the Chakra's Accordion component which makes this simpler and cleaner.
```suggestion
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+              flexDir={{ base: "column", sm: "row" }}
+            >
+              <Box
+                marginBottom={{ base: "0.5rem", sm: 0 }}
+                marginRight={{ base: 0, sm: "1rem" }}
```
```suggestion
                marginBottom={{ base: 2, sm: 0 }}
                marginRight={{ base: 0, sm: 4 }}
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+                  sx={{
+                    svg: { marginRight: "1.5rem" },
+                  }}
+                  margin="1rem 0"
```
```suggestion
                  my="4"
```

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +36,121 @@ const ExpandableCard: React.FC<IProps> = ({
+            <Box
+              fontSize="md"
+              color="text"
+              marginTop="2rem"
+              paddingTop="1.5rem"
```
```suggestion
              marginTop="8"
              paddingTop="6"
```

pettinarip reviewed (changes requested):
Good job @jb-0. Left a couple of comments to simplify a bit the implementation & to be consistent with the codebase style.

pettinarip commented on src/components/ExpandableCard.tsx:
```diff
@@ -107,111 +34,114 @@ const ExpandableCard: React.FC<IProps> = ({
+          <Box
+            fontSize="md"
+            color="text"
+            marginTop="8"
```
```suggestion
```

pettinarip reviewed (approved):
lgtm! just removed an extra margin to make it look a bit nicer. Good job @jb-0

jb-0 commented:
> lgtm! just removed an extra margin to make it look a bit nicer. Good job @jb-0 

Nice one, thanks for your help!
